CDAP-17981: Make FileLocalizer to load files via AppFab instead of di…

wyzhang opened the pull request:
…rectly from HDFS

Why:
FileLocalizer is used to localizing files needed to run twill application in k8s env.
The file is most likely in an external distributed file system (e.g. GCS in GCP)
The pod to run twill application may or may not have permission to access distributed
file system. So this change swtiches the localization to downloading from AppFab
(essentially make AppFab proxy the file content to twill application pod)

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,109 @@
+ */
+@Singleton
+@Path(Constants.Gateway.INTERNAL_API_VERSION_3)
+public class FileFetcherHttpHandlerInternal extends AbstractAppFabricHttpHandler {
```
Do you need something specific from the AbstractAppFabricHttpHandler class? If not, better not to extend from it.

wyzhang replied:
Right. Fixed. Replace with AbstractHttpHandler

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,109 @@
+  private final LocationFactory locationFactory;
+
+  @Inject
+  FileFetcherHttpHandlerInternal(CConfiguration configuration,
```
The cConf is unused.

wyzhang replied:
Removed.

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,109 @@
+  @GET
+  @Path("/location/{encodedUri}")
+  public void download(HttpRequest request, HttpResponder responder,
+                          @PathParam("encodedUri") String encodedUri) throws Exception {
```
Misalign.

wyzhang replied:
Fixed. Reformated the entire file. Need to remember doing that for all files touched.

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  /**
+   * Download the file specified by the base64-encoded URI
```
Why base64 encode the URI? Shouldn't the file be always from the same base as the LocationFactory (i.e we shouldn't allow remote call to fetch files on the local fs of app-fabric).

wyzhang replied:
Without base64 encoding, the URL path becomes v3Internal/location/a/b/c/e/d, which results in not found, as file path gets treated as part of  the URL. 

Added a check in the http handler to ensure the file scheme is the same in the incoming URI and location factory.

tivv replied:
Base64 is still pretty unconventional in HTTP world. Can we do it as parameter and use regular URL encoding?

chtyim replied:
You can use `**` to match any suffix. See https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeServiceRoutingHandler.java#L104 on how to use it.

tivv replied:
I was actually looking at it but was not sure you will get the whole thing into the last path variable

wyzhang replied:
Done. Fixed

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,109 @@
+        public void handleError(@Nullable Throwable cause) {
+          Closeables.closeQuietly(inputStream);
+        }
+      }, EmptyHttpHeaders.INSTANCE);
```
Shouldn't you set the content-type?

wyzhang replied:
Added.

For my own edification purpose, the test passes without setting the header. So I suppose for inter cluster communication, netty http lib doesn't check and require content type header to be set?

chtyim replied:
There is no hard requirement in HTTP that content type has to be set. But it is always good to set.

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,109 @@
+        }
+      }, EmptyHttpHeaders.INSTANCE);
+    } catch (Exception e) {
+      Closeables.closeQuietly(inputStream);
```
Move this to use `try (InputStream is = location.getInputStream()) {` instead.

wyzhang replied:
streaming the file content in chunks appears to run asyncly. If using try with resource, the input stream gets closed almost immediately before responder gets a chance to read and send any chunk asyncly.

tivv replied:
Well, it should not behave differently if you do in same "try". It will close the stream on try exit. If it's async, you will get same problems with current approach as well

chtyim replied:
Actually you are right @wyzhang . Keep it as it since we only close it if there is exception raised from the sendContent call.

wyzhang replied:
Try and catch don't close the stream for you upon existing try clause, while try-with-resource does. 

Confirmed that test fails with try-with-resource, while passes with current approach.

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/master/environment/MasterEnvironments.java:
```diff
@@ -24,9 +24,12 @@
 import io.cdap.cdap.common.conf.CConfiguration;
 import io.cdap.cdap.common.guice.ConfigModule;
 import io.cdap.cdap.common.guice.DFSLocationModule;
+import io.cdap.cdap.common.guice.SupplierProviderBridge;
```
Unused imports??

wyzhang replied:
Fixed.

chtyim commented on cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/FileFetcherHttpHandlerInternalTest.java:
```diff
@@ -0,0 +1,125 @@
+/**
+ * Test for {@link FileFetcherHttpHandlerInternal}.
+ */
+public class FileFetcherHttpHandlerInternalTest extends AppFabricTestBase {
```
I would suggest not to extend from AppFabricTestBase. It is a heavy class that starts a lot of services. To just test the handler, create a `NettyHttpService` with the handler you wanted to test.

CuriousVini replied:
+1 similar to 
https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeServiceRoutingTest.java#L139

wyzhang replied:
Agreed. Done.

chtyim commented on cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/FileFetcherHttpHandlerInternalTest.java:
```diff
@@ -0,0 +1,125 @@
+    HttpRequest request = remoteClient.requestBuilder(
+      HttpMethod.GET,
+      String.format("location/%s", Base64.getEncoder().encodeToString(location.toURI().toString().getBytes())))
+      .withContentConsumer(
```
For testing purpose, it could be simplified to let the remoteClient to just fetch everything in memory. Alternatively, create a `FileContentConsumer` that saves the content to a file (which can be shared with the `HubPackage` class).

wyzhang replied:
Got rid of remote client. PTAL.

chtyim commented on cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/FileFetcherHttpHandlerInternalTest.java:
```diff
@@ -0,0 +1,125 @@
+    // Verify the source and destination files are identical.
+    byte[] orgFileContent = Files.readAllBytes(orgFile.toPath());
+    byte[] downloadedFileContent = Files.readAllBytes(targetFile.toPath());
+    Assert.assertEquals(orgFileContent.length, downloadedFileContent.length);
```
Just use `Assert.assertArrayEquals()` to unify to one assert call.

wyzhang replied:
Done. Fixed. 

I broke this up into length check and content check since it facilitated my debugging, as my earlier implementation incorrectly wrote the entire buffer to file, instead of the portion of the buffer containing data. The length check better surfaces that issue.

CuriousVini commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,109 @@
+@Singleton
+@Path(Constants.Gateway.INTERNAL_API_VERSION_3)
+public class FileFetcherHttpHandlerInternal extends AbstractAppFabricHttpHandler {
+  private static final Gson GSON = new Gson();
```
nit: unused?

wyzhang replied:
Removed.

CuriousVini commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,109 @@
+   * @param request   {@link HttpRequest}
+   * @param responder {@link HttpResponse}
+   * @param encodedUri base64 encoded URI that identifies the target file to download.
+   * @throws Exception if the file is not found or downloading fails.
```
better return specific exceptions for the caller such as filenotfound Or bad request etc

wyzhang replied:
Ended up using responder.sendString. When using BadRequest, it somehow gets translates to internal errors on the caller side.

CuriousVini commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,109 @@
+
+        @Override
+        public void handleError(@Nullable Throwable cause) {
+          Closeables.closeQuietly(inputStream);
```
Should we log some error in case of error?

wyzhang replied:
Good point. Done.

tivv replied:
At the same time, is not warn a big high? network errors are not that rare. Would it flood log with "BrokenPipe" exceptions?

wyzhang replied:
Changed to debug.

CuriousVini commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileFetcher.java:
```diff
@@ -0,0 +1,99 @@
+ */
+public class FileFetcher {
+  private static final Logger LOG = LoggerFactory.getLogger(FileFetcher.class);
+  private RemoteClient remoteClient = null;
```
nit: no need to initialize null

wyzhang replied:
Removed

CuriousVini commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileLocalizer.java:
```diff
@@ -49,10 +48,12 @@
 
-  public FileLocalizer(MasterEnvironmentContext context, @SuppressWarnings("unused") MasterEnvironment masterEnv) {
+  public FileLocalizer(MasterEnvironmentContext context, MasterEnvironment masterEnv) {
     this.context = context;
```
`MasterEnvironmentContext` is not being used?

CuriousVini commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,109 @@
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.io.Locations;
+import io.cdap.cdap.gateway.handlers.util.AbstractAppFabricHttpHandler;
+import io.cdap.cdap.proto.ApplicationDetail;
```
nit: unused

CuriousVini commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,109 @@
+/**
+ * Internal {@link HttpHandler} for serving file downloading requests.
+ *
+ * This is currently used by FileLocalizer to download required files from AppFabric
```
good to add class with `{@link FileLocalizer}`

CuriousVini commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileFetcher.java:
```diff
@@ -0,0 +1,99 @@
+ * The target file could be either residing in AppFabric's local file system or a distributed file system
+ * that AppFabric is configured to access.
+ */
+public class FileFetcher {
```
If this is just being used in this package we can make this class and `download` method package private?

wyzhang replied:
Yes. Done.

CuriousVini commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileFetcher.java:
```diff
@@ -0,0 +1,99 @@
+          outputStream.write(bytes);
+          outputStream.flush();
+        } catch (IOException e) {
+          LOG.error("Failed to download file from {} to {}", sourceURI, targetLocation.toURI());
```
should include exception in the logs

wyzhang replied:
Added.

CuriousVini commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileFetcher.java:
```diff
@@ -0,0 +1,99 @@
+      @Override
+      public void onFinished() {
+        try {
+        outputStream.close();
```
May be use ` Closeables.closeQuietly(outputStream);`

wyzhang replied:
Tried that, but it takes only inputstream or reader.

CuriousVini commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileFetcher.java:
```diff
@@ -0,0 +1,99 @@
+    httpResponse.consumeContent();
+
+    if (httpResponse.getResponseCode() != HttpResponseStatus.OK.code()) {
+      throw new IOException("Failed to download file from {} to {} " + httpResponse.getResponseBodyAsString());
```
should we return different exceptions for the caller rather than wrapping all in `IOException`?

wyzhang replied:
Good point. Updated. PTAL

CuriousVini commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileFetcher.java:
```diff
@@ -0,0 +1,99 @@
+      }
+    }).build();
+    HttpResponse httpResponse = HttpRequests.execute(request, new DefaultHttpRequestConfig(false));
+    httpResponse.consumeContent();
```
could we use `HttpClients#executeStreamingRequest` instead? That method abstracts out retries as well

wyzhang replied:
That would add cdap-app-fabric as new dependency.  Not sure if that is concern, thus keeping this as is. But could look into refactoring that out to share the code in a follow up PR.

chtyim commented on cdap-kubernetes/pom.xml:
```diff
@@ -86,6 +86,12 @@
     </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-common</artifactId>
```
Why? the kube module shouldn't have dependency on common.

wyzhang replied:
RemoteClient (i.e. libs to build http request and make REST API calls) brings in the dependency. 

To avoid this dependency, would need to duplicate some code over. Let me know if you see any better alternative.

chtyim replied:
First of all, the kube module (or any master-spi extension for that matter), is designed with classloader isolation, so adding dependency on cdap-common means it will bring a different copy (and all of its transitive dependencies) into the extension directory. The cdap-common traditionally is a dump place for internal classes that needs to be shared among multiple modules, which makes it kind of big and full of weird dependencies.

If it is just for downloading a file from a http endpoint purpose, using the URLConnection directly doesn't really make the code more complicated (in fact might be easier).

wyzhang replied:
Reverted. 

Moved to MasterEnvironment. Tried to move to MasterEnvironmentContext, but no disocveryService there (which gets created after context creation)

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileFetcher.java:
```diff
@@ -0,0 +1,106 @@
+            byte[] bytes = new byte[chunk.remaining()];
+            chunk.get(bytes, 0, bytes.length);
+            outputStream.write(bytes);
+            outputStream.flush();
```
Why need to flush?

wyzhang replied:
Flush so inmemory buffer can be freed (hoptefully), as an optimization.

tivv replied:
The buffers are rarely freed on flush, pretty often those are ringbuffers or reusable buffers. But flush may require a drive sync or network call

tivv replied:
if you want some optimization, your best bet is to check `chunk.hasArray` and use it's array if available. This would save you a buffer allocation and one copy

tivv replied:
Or actually best way is to use channels, as seen here: https://stackoverflow.com/questions/579600/how-to-put-the-content-of-a-bytebuffer-into-an-outputstream
```
public void writeBuffer(ByteBuffer buffer, OutputStream stream) {
   WritableByteChannel channel = Channels.newChannel(stream);

   channel.write(buffer);
}```

chtyim replied:
Which in-memory buffer you are talking about? Internal to the FileOutputStream or the `bytes` you have? In either case, calling flush doesn't free anything. Depending on the implementation, flush at the best is to call `fsync` to force data to get written out to the storage device. However, memory usage is still the same after the flush because memory is not freed, but rather emptied. In this particular case, since we only use the file after it is closed, flushing on every write would only make the download slower, without any benefits. (flush/sync is useful when doing write-ahead-log that type of logic to prevent data loss on crashing).

wyzhang replied:
Original intention is to flush so the new byte[] can be freed right away. But it does depend on implementation. 
The concern I had was high memory usage due to too many new bytes[] calls across multiple iterations that underlying library holds onto them. 

Anyway, removed the flush

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileFetcher.java:
```diff
@@ -0,0 +1,106 @@
+          try {
+            outputStream.close();
+          } catch (Exception e) {
+            LOG.error("Failed to close the file downloaded from {} to {}", sourceURI, targetLocation.toURI());
```
Shouldn't it fail as well? The download should be considered failed if failing to close the stream.

wyzhang replied:
Good catch. Moved the outpustream.close() out

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileFetcher.java:
```diff
@@ -0,0 +1,106 @@
+    if (httpResponse.getResponseCode() != HttpResponseStatus.OK.code()) {
+      if (httpResponse.getResponseCode() == HttpResponseStatus.NOT_FOUND.code()) {
+        throw new FileNotFoundException(httpResponse.getResponseBodyAsString());
+      } else if (httpResponse.getResponseCode() == HttpResponseStatus.BAD_REQUEST.code()) {
```
No need to have else after the if block throw. It is cleaner to write:

```java
if (...) {
  throw ...;
}
if (...) {
  throw ...
}
```

wyzhang replied:
Fixed.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileFetcher.java:
```diff
@@ -0,0 +1,103 @@
+   * @param targetLocation target location to store the downloaded file
+   * @throws IOException if file downloading or writing to target location fails.
+   */
+  void download(URI sourceURI, Location targetLocation) throws IOException, BadRequestException {
```
Why the target is a `Location`? Shouldn't it always be `File` or `Path`?

wyzhang replied:
Fixed.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileLocalizer.java:
```diff
@@ -86,13 +87,15 @@ public void run(String[] args) throws Exception {
+          expand(localizedFile, targetPath);
         } else {
-          copy(location, targetPath);
+          copy(localizedFile, targetPath);
```
Shouldn't copy again. The file has been downloaded to local disk already. You should either link it, or better yet having the file fetcher write to this destination directly.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileLocalizer.java:
```diff
@@ -86,13 +87,15 @@ public void run(String[] args) throws Exception {
 
         if (localFile.isArchive()) {
-          expand(location, targetPath);
+          expand(localizedFile, targetPath);
```
In here, you actually can start expanding the file when downloading the bytes (via ZipInputStream). This means the download should probably be as `download(URI source, OutputStream os)`.

chtyim commented on cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/FileFetcherHttpHandlerInternalTest.java:
```diff
@@ -0,0 +1,160 @@
+            byte[] bytes = new byte[chunk.remaining()];
+            chunk.get(bytes, 0, bytes.length);
+            outputStream.write(bytes);
+            outputStream.flush();
```
Shouldn't need to flush

wyzhang replied:
Done

chtyim commented on cdap-app-fabric-tests/src/test/java/io/cdap/cdap/internal/app/runtime/distributed/ProgramTwillRunnableModuleTest.java:
```diff
@@ -178,5 +181,10 @@ public String getName() {
+
+    @Override
+    public void fetchFile(URI srcURI, File dest) throws IOException {
+      throw new IOException("Unimplemented");
```
Should throw `UnsupportedOperationException`

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileLocalizer.java:
```diff
@@ -64,17 +65,16 @@ public void run(String[] args) throws Exception {
-    } else {
-      runtimeConfigLocation = new LocalLocationFactory().create(new File(uri).toURI());
-    }
+    Location runtimeConfigLocation = new LocalLocationFactory().create(new File(uri.getPath()).toURI());
```
Why the above if check is removed? If the `uri` is already referring to a local file, it shouldn't need to fetch from the masterEnv?

wyzhang replied:
Revived the code 

    if (context.getLocationFactory().getHomeLocation().toURI().getScheme().equals(uri.getScheme())) {
      runtimeConfigLocation = context.getLocationFactory().create(uri);

The code only checks LocationFactory and uri share the same filesystem scheme. They could both refer to the same non-local file system, right?

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileLocalizer.java:
```diff
@@ -64,17 +65,16 @@ public void run(String[] args) throws Exception {
-    }
+    Location runtimeConfigLocation = new LocalLocationFactory().create(new File(uri.getPath()).toURI());
+
+    masterEnv.fetchFile(uri, new File(runtimeConfigLocation.toURI().getPath()));
```
Shouldn't the fetch method should be in the context instead of environment? The `MasterEnvironmentContext` is for CDAP to provide core services to the `MasterEnvironment` implementations. The `MasterEnvironment` is for CDAP to call into the extension implementation.

wyzhang replied:
As mentioned above, initiated tried to do in context, but context created before master env creation, where discovery service gets created.

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileLocalizer.java:
```diff
@@ -64,17 +65,16 @@ public void run(String[] args) throws Exception {
-    try (Reader reader = Files.newBufferedReader(runtimeConfigDir.resolve(Constants.Files.TWILL_SPEC),
-                                                 StandardCharsets.UTF_8)) {
+    try (Reader reader = Files.newBufferedReader(
+      runtimeConfigDir.resolve(org.apache.twill.internal.Constants.Files.TWILL_SPEC),
```
Why need to use long class name?

wyzhang replied:
Fixed. Used to conflict with constants from cdap-common package.

wyzhang committed:
CDAP-17981: Make FileLocalizer to load files via AppFab instead of directly from HDFS

Why:
FileLocalizer is used to localizing files needed to run twill application in k8s env.
The file is most likely in an external distributed file system (e.g. GCS in GCP)
The pod to run twill application may or may not have permission to access distributed
file system. So this change swtiches the localization to downloading from AppFab
(essentially make AppFab proxy the file content to twill application pod)

wyzhang committed:
cleanup

wyzhang committed:
some more refactoring

wyzhang commented:
Discard this PR. Have taken a different approach to implement this based on offline discussion, which have already been submitted by now.
